### PR TITLE
506 account creation, token verification and refresh (service only)

### PR DESCRIPTION
**Pull request by clemens-msupply**
Plan to implement logout while I am on it in a different PR

I used a UserAccountService struct since I plan to have a "TokenBucket" for logout which I like to pass in when creating the UserAccountService...

### CODE REVIEW

File: src/service/user_account.rs [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+                JWTValidationError::ExpiredSignature
+            }
+            jsonwebtoken::errors::ErrorKind::InvalidAudience => JWTValidationError::NotAnApiToken,
+            _ => JWTValidationError::InvalidToken(err),
+        })?;
+
+        Ok(decoded.claims)
+    }
+
+    pub fn refresh_token(
```
Comment (andreievg) [inline src/service/user_account.rs:186]: does this method need to be in struct ?

File: src/service/user_account.rs [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+                    return Err(JWTIssuingError::InvalidCredentials(password_result));
+                }
+            }
+            _ => return Err(JWTIssuingError::InvalidCredentials(password_result)),
+        }
+
+        create_jwt_pair(&user.id, valid_for).map_err(|err| JWTIssuingError::CanNotCreateToken(err))
+    }
+
+    pub fn verify_token(&self, token: &str) -> Result<OmSupplyClaim, JWTValidationError> {
```
Comment (andreievg) [inline src/service/user_account.rs:166]: does this method need to be in struct ?

**Review by andreievg — COMMENTED** [review]
Thanks overall looks good, a few questions and suggestions. Also not in the review comments:

Personally i would create a common verify jwt, to which audience is passed, and use it in refresh_token.

File: src/service/user_account.rs [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+    ) -> Result<TokenPair, JWTIssuingError> {
+        let repo = UserAccountRepository::new(self.connection);
+        let user = repo
+            .find_one_by_user_name(username)
+            .map_err(|err| match err {
+                RepositoryError::NotFound => JWTIssuingError::UserNameDoesNotExist,
+                _ => JWTIssuingError::DatabaseError(err),
+            })?;
+        // verify password
+        let password_result = verify(password, &user.password);
```
[the remark this answers is not in the document]
Comment (clemens-msupply) [reply src/service/user_account.rs]: done in #ad1d6d8

File: src/service/user_account.rs [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+        match password_result {
+            Ok(valid) => {
+                if !valid {
+                    return Err(JWTIssuingError::InvalidCredentials(password_result));
+                }
+            }
+            _ => return Err(JWTIssuingError::InvalidCredentials(password_result)),
```
[the remark this answers is not in the document]
Comment (clemens-msupply) [reply src/service/user_account.rs]: done in #ad1d6d8

File: src/service/user_account.rs [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+    fn from(err: RepositoryError) -> Self {
+        CreateUserAccountError::DatabaseError(err)
+    }
+}
+
+/// Error for getting a JWT token
+#[derive(Debug)]
+pub enum JWTIssuingError {
+    UserNameDoesNotExist,
+    InvalidCredentials(Result<bool, bcrypt::BcryptError>),
```
[the remark this answers is not in the document]
Comment (clemens-msupply) [reply src/service/user_account.rs]: done in ad1d6d8

File: src/service/user_account.rs (hunk 1 above)
Comment (clemens-msupply) [reply src/service/user_account.rs:191]: Think in this case its nice to have it together with the other methods which need access to the repo but its also needed to be in there for the logout PR.

File: src/service/user_account.rs (hunk 2 above)
Comment (clemens-msupply) [reply src/service/user_account.rs:168]: as above

File: src/service/user_account.rs [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+    };
+    let api_token = jsonwebtoken::encode(
+        &jsonwebtoken::Header::default(),
+        &api_claims,
+        &jsonwebtoken::EncodingKey::from_secret(JWT_TOKEN_SECRET),
+    )?;
+
+    // refresh token
+    let refresh_claims = OmSupplyClaim {
+        exp: now + valid_for,
```
[the remark this answers is not in the document]
Comment (clemens-msupply) [reply src/service/user_account.rs]: add option to specify refresh token expiry in 64c39e2

File: src/service/user_account.rs [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+    ExpiredSignature,
+    NotAnApiToken,
+    InvalidToken(jsonwebtoken::errors::Error),
+}
+
+#[derive(Debug)]
+pub enum JWTRefreshError {
+    ExpiredSignature,
+    NotARefreshToken,
+    InvalidToken(jsonwebtoken::errors::Error),
```
[the remark this answers is not in the document]
Comment (clemens-msupply) [reply src/service/user_account.rs]: done 42decf1

**Comment by clemens-msupply** [comment]
Thanks for the review. Regarding the token, refresh token creation, I decided to leave it as it for now since the method is relatively small and find it a bit nicer to have it all in one method, not strong about it though.

**Review by andreievg — APPROVED** [review]
Thanks for the changes, looks good

**Pull request merged**